Double-precision build fails

Our project (https://www.skyscrapersim.net) is probably one of the few that uses Ogre in double-precision mode, in order to support very large environments, eventually we could probably move to a different method, but for now this is what we use.

Originally users reported a large drop in FPS when upgrading from Ogre 13 to 14.0.1, but the latest 14 appears to run better, so that might no longer be an issue.  Currently I'm getting a number of build errors when building both the released 14.1.2 and the latest Git code of Ogre in double-precision mode, such as:
`/data/src/ogre-dev/OgreMain/src/OgreAutoParamDataSource.cpp: In member function ‘Ogre::Vector4f Ogre::AutoParamDataSource::getSpotlightParams(size_t) const’:
/data/src/ogre-dev/OgreMain/src/OgreAutoParamDataSource.cpp:243:33: error: no match for ‘operator*’ (operand types are ‘Ogre::Matrix3’ and ‘Ogre::Vector3f’ {aka ‘Ogre::Vector<3, float>’})
`

I tried to fix a number of them by re-casting as floats, but decided to open a bug report about it instead.  Thanks

--- a/Components/RTShaderSystem/src/OgreShaderExIntegratedPSSM3.h
+++ b/Components/RTShaderSystem/src/OgreShaderExIntegratedPSSM3.h
@@ -49,7 +49,7 @@ class IntegratedPSSM3 : public SubRenderState
 
     // Interface.
 public:
-    typedef std::vector<Real> SplitPointList;
+    typedef std::vector<float> SplitPointList;
 
     /** Class default constructor */    
     IntegratedPSSM3();

--- a/OgreMain/src/OgreAutoParamDataSource.cpp
+++ b/OgreMain/src/OgreAutoParamDataSource.cpp
@@ -220,7 +220,7 @@ namespace Ogre {
         if(l.getType() == Light::LT_RECTLIGHT)
         {
             auto rot = getViewMatrix().linear();
-            return Vector4f(rot * l.getDerivedSourceHalfHeight(), 0.0);
+            return Vector4f(Vector3f(rot * Vector3(l.getDerivedSourceHalfHeight())), 0.0);
         }
         // range, const, linear, quad
         return l.getAttenuation();
@@ -240,7 +240,7 @@ namespace Ogre {
         else if(l.getType() == Light::LT_RECTLIGHT)
         {
             auto rot = getViewMatrix().linear();
-            return Vector4f(rot * l.getDerivedSourceHalfWidth(), 2.0);
+            return Vector4f(Vector3f(rot * Vector3(l.getDerivedSourceHalfWidth())), 2.0);
         }
         else
         {

--- a/OgreMain/src/OgreGpuProgramParams.cpp
+++ b/OgreMain/src/OgreGpuProgramParams.cpp
@@ -1204,7 +1204,7 @@ namespace Ogre
         size_t logicalIndex, size_t requestedSize, uint16 variability, BaseConstantType type)
     {
         OgreAssert(type != BCT_SAMPLER, "");
-        if (!mLogicalToPhysical)
+        if (!mLogicalToPhysical || int(logicalIndex) == -1)
             return NULL;
 
         GpuLogicalIndexUse* indexUse = 0;
@@ -1229,8 +1229,7 @@ namespace Ogre
 
                 // Set up a mapping for all items in the count
                 size_t currPhys = physicalIndex;
-                size_t count = requestedSize / 4;
-                GpuLogicalIndexUseMap::iterator insertedIterator;
+                size_t count = std::max<size_t>(requestedSize / 4, 1);
 
                 for (size_t logicalNum = 0; logicalNum < count; ++logicalNum)
                 {
@@ -1241,10 +1240,8 @@ namespace Ogre
                     currPhys += 4;
 
                     if (logicalNum == 0)
-                        insertedIterator = it;
+                        indexUse = &(it->second);
                 }
-
-                indexUse = &(insertedIterator->second);
             }
             else
             {
@@ -1255,15 +1252,15 @@ namespace Ogre
         }
         else
         {
-            size_t physicalIndex = logi->second.physicalIndex;
             indexUse = &(logi->second);
             // check size
-            if (logi->second.currentSize < requestedSize)
+            if (indexUse->currentSize < requestedSize)
             {
+                size_t physicalIndex = indexUse->physicalIndex;
                 // init buffer entry wasn't big enough; could be a mistake on the part
                 // of the original use, or perhaps a variable length we can't predict
                 // until first actual runtime use e.g. world matrix array
-                size_t insertCount = requestedSize - logi->second.currentSize;
+                size_t insertCount = requestedSize - indexUse->currentSize;
                 auto insertPos = mConstants.begin();
                 std::advance(insertPos, physicalIndex);
                 mConstants.insert(insertPos, insertCount*4, 0);
@@ -1293,7 +1290,7 @@ namespace Ogre
                     mNamedConstants->bufferSize += insertCount;
                 }
 
-                logi->second.currentSize += insertCount;
+                indexUse->currentSize += insertCount;
             }
         }
 

--- a/OgreMain/src/OgreLight.cpp
+++ b/OgreMain/src/OgreLight.cpp
@@ -295,13 +295,13 @@ namespace Ogre {
     {
         auto& ori = mParentNode->_getDerivedOrientation();
         auto& scale = mParentNode->_getDerivedScale();
-        return ori.xAxis() * mSourceSize[0] * scale[0] * 0.5f;
+        return Vector3f(ori.xAxis()) * mSourceSize[0] * scale[0] * 0.5f;
     }
     Vector3f Light::getDerivedSourceHalfHeight() const
     {
         auto& ori = mParentNode->_getDerivedOrientation();
         auto& scale = mParentNode->_getDerivedScale();
-        return ori.yAxis() * mSourceSize[1] * scale[1] * 0.5f;
+        return Vector3f(ori.yAxis()) * mSourceSize[1] * scale[1] * 0.5f;
     }
     //-----------------------------------------------------------------------
     const PlaneBoundedVolume& Light::_getNearClipVolume(const Camera* const cam) const

--- a/RenderSystems/GL3Plus/src/GLSL/src/OgreGLSLShader.cpp
+++ b/RenderSystems/GL3Plus/src/GLSL/src/OgreGLSLShader.cpp
@@ -574,11 +574,11 @@ namespace Ogre {
             if (name == "OgreUniforms") // default buffer
             {
                 extractUniforms(blockIdx);
-                int binding = int(mType);
+                int binding = mType == GPT_COMPUTE_PROGRAM ? 0 : int(mType);
                 if (binding > 1)
                     LogManager::getSingleton().logWarning(
                         getResourceLogName() +
-                        " - using a UBO in this shader type will alias with shared_params");
+                        " - using 'OgreUniforms' in this shader type does alias with shared_params");
 
                 mDefaultBuffer = hbm.createUniformBuffer(values[2]);
                 static_cast<GL3PlusHardwareBuffer*>(mDefaultBuffer.get())->setGLBufferBinding(binding);

